### ISSUE TITLE
Firefox: using a link to get to Cockpit login page results in 401 Auth failed

### ISSUE THREAD

**Issue by Aerowinder** [cockpit-project/cockpit#17936]
### Explain what happens

AlmaLinux 9.1, Cockpit 276.1-1, Firefox 107

Using the homelab dashboard "Dashy". When I click a link in Dashy, a new Firefox tab is opened to the target destination of the Cockpit system. I'm going from my domain (reverse proxy) to the IPv4 of the Cockpit system (but IP to IP exhibits same issue).

Attempting to login to Cockpit this way produces the issue where the credentials aren't declined (on screen, like if you enter an incorrect password), but the login page just reloads. Firefox reports "Cookie “cockpit” with the “SameSite” attribute value “Lax” or “Strict” was omitted because of a cross-site redirect." But there doesn't appear to be any redirect happening with Dashy.

This setup works on Edge/Chromium (can follow dashboard link to Cockpit) with no changes.
This setup works with every other web services I'm hosting.

This does not appear to be a Dashy problem, as it works with every other web service I'm using, and works with Cockpit if using Edge/Chromium.
This does not appear to be a Firefox problem, as it works with every other web service I'm using. Additionally, if I copy/paste the Cockpit address in the address bar, I am able to login normally.

I'm not sure what the problem is, hoping somebody here will be able to point me in the right direction.

Only Cockpit seems to be struggling here, and I can't find any information on this issue. I've attempted implementing sections from https://cockpit-project.org/guide/latest/cockpit.conf.5, but that hasn't helped. Couldn't find any errors in the Cockpit logs (journalctl -u cockpit)

### Version of Cockpit

276.1

### Server operating system

Red Hat Enterprise Linux

### Server operating system version

9.1

### What browsers are you using?

Firefox, Edge

**Comment by garrett** [comment]
This isn't a specifically a Firefox + Cockpit issue either, as most of the team uses Firefox for development. We also actively test using CentOS Stream and RHEL, so this probably isn't an issue with Alma either, as it's a rebuild of RHEL.

> I'm going from my domain (reverse proxy) to the IPv4 of the Cockpit system (but IP to IP exhibits same issue).

I know that people sometimes have issues with proxies; this is the most likely culprit. But if you mean a direct connection still has the issue, I guess it's not really the proxy itself?

All of this makes me really suspicious of your certificate. Perhaps that's the issue? Firefox and Chrome use different certificate stores and accept different types of certificates.

---

Meanwhile, I did look up the SameSite behavior and Firefox specifically and found a post @ https://hacks.mozilla.org/2020/08/changes-to-samesite-cookie-behavior/

It does link a testing sandbox site @ https://samesite-sandbox.glitch.me/, which shows this for me:

![sandbox results; there are six errors, one invalid, and eleven are green](https://user-images.githubusercontent.com/10246/205102100-df639c6f-6e99-46ba-988b-92b0650b25dc.png)

However, they're all green in Firefox Nightly for me:

![image](https://user-images.githubusercontent.com/10246/205102562-573a195f-2c65-44b1-b19e-22c0c9a369a6.png)

It might be the case that a privacy add-on that I have in my main Firefox browser, such as uBlock Origin, or LocalCDN could be causing issues for me in the sandbox testcase. (However, this doesn't cause an issue for me when using Cockpit locally.)

It could also be the case that an add-on could be causing you issues with Firefox talking to your Cockpit installation too.

1. What does the sandbox page show for you?
2. Could you test Firefox in safe mode? (Or install a version of Firefox, like Beta/Dev/Nightly in parallel and see what happens in a fresh install.)

To run Firefox in safe mode, go to the menu, select "Help" and then choose "Troubleshoot mode...". It will show a modal dialog that asks if you'd like to restart Firefox in a troubleshooting mode and it will mention that extensions, themes, and custom settings will be temporarily disabled.

Firefox will restart and then you can check what happens with a fresh installation. When you're done, restart Firefox and you should be back to normal.

---

When you restart, you'll lose your tabs and tab history. I'm not sure if Firefox will let you restore them or not. If that doesn't work for you, then you could install [Firefox Developer Edition](https://www.mozilla.org/en-US/firefox/developer/), which is a standalone version of Firefox, meaning you can run both at the same time.

If you're on Linux (it's not clear, as Edge is also available for Linux), then there's also [Floorp](https://flathub.org/apps/details/one.ablaze.floorp) and [LibreWolf](https://flathub.org/apps/details/io.gitlab.librewolf-community) ­— both are Firefox forks and available on Flathub, so they're easy to install and run in parallel. (You can also run Firefox from Mozilla's tarball, but you'd have to do so from the command line.)

**Comment by garrett** [comment]
OK, that was a lot. Whoops. Here's a summary:

- Certificate, perhaps?
- Firefox policy change? (Probably not.)
- Add-on in Firefox?
- To see if it's an add-on for Firefox or a configuration setting that you might've changed (even unknowingly):
   - you can try safe mode
   - use a different version of Firefox (such as Developer Edition) if you'd like to run it at the same time as your main Firefox.
- Please check https://samesite-sandbox.glitch.me/

**Comment by CoolMatty** [comment]
I am also having extremely similar behavior. Firefox latest stable, addons disabled, and I get returned to the login page with no visible error. Dev console reports the same cross-site cookie error:

`Cookie “cockpit” with the “SameSite” attribute value “Lax” or “Strict” was omitted because of a cross-site redirect.`

Adjusting cookie/security settings for the page had no effect.

However, I am running a local self-signed cert for my Cockpit install. I should note that this worked fine as of a month ago.

Finally, I also noticed that if I directly access Cockpit by IP and not by its domain name, the login issue goes away. 

For sanity, I also checked with Microsoft Edge, and that was able to login fine regardless of method.

**Comment by Aerowinder** [comment]
It certainly sounds similar. The difference is I can go from (dashy > cockpit) domain > IP or IP > IP and both fail. If I put cockpit in my reverse proxy (using my LE cert), it works as expected.

Right now though, I am using the self-signed cert cockpit generated.

**Comment by martinpitt** [comment]
> Firefox reports "Cookie “cockpit” with the “SameSite” attribute value “Lax” or “Strict” was omitted because of a cross-site redirect

That's almost surely the root cause. cockpit's web server does set `SameSite=Strict` for its cookie, and that somehow seems to make Firefox omit it when redirecting from http://10.10.3.5:4000 to https://10.10.3.1:9090/. So indeed this should work *better* with a reverse proxy, as then both dashy and cockpit come from the same origin. That also makes it easier to get proper TLS certificates, as cockpit then doesn't have to have its own. Is that an option for you?

> it works with every other web service I'm using

Supposedly none of them set the "SameSite" cookie attribute? 

> Are there any other logs I could provide other than the 401 Firefox gives me in the dev console? 

I'm afraid not. This isn't visible on the server side, it's a policy that the browser itself decides.

**Comment by Aerowinder** [comment]
> That's almost surely the root cause. cockpit's web server does set `SameSite=Strict` for its cookie, and that somehow seems to make Firefox omit it when redirecting from http://10.10.3.5:4000 to https://10.10.3.1:9090/. So indeed this should work _better_ with a reverse proxy, as then both dashy and cockpit come from the same origin. That also makes it easier to get proper TLS certificates, as cockpit then doesn't have to have its own. Is that an option for you?

Yes, I am able to get all of my Cockpit instances into a reverse proxy. And it works great with that setup.

> Supposedly none of them set the "SameSite" cookie attribute?

I guess not. I'm not sure where to check this setting.

> I'm afraid not. This isn't visible on the server side, it's a policy that the browser itself decides.

Understood, thanks for the assistance.


Appreciate the help all, I'll mark this as closed/wontfix since this behavior is intended for security reasons.

**Comment by Br0kenSilos** [comment]
Just ran into this as well.  In my case, I am connecting to Cockpit through a CloudFlare Tunnel. If I click the link in Dashy, the site opens in a new tab as expected. But I am unable to login. If I then go up to the address bar and just hit enter, the login then works fine.

**Comment by garrett** [comment]
There's an older issue related to CloudFlare Argo, which is the previous name of CloudFlare Tunnel, at https://github.com/cockpit-project/cockpit/issues/16396 which mentions configuration changes.

Could you look at that and see if that fixes your issue? Come back here to say if it worked or not.

And then, this might also be an issue with CloudFlare Tunnel or Dashy. 

There is a plugin for Cockpit related to CloudFlare Tunnels @ https://github.com/spotsnel/cockpit-cloudflared, so that means some people are successfully using CloudFlare Tunnel with Cockpit. So the problem is likely with Dashy, which is mentioned higher up in this issue (without mentioning CloudFlare). If you can't get things working, perhaps open an issue up @ https://github.com/Lissy93/dashy/issues/ with a link back to this issue?

Also, I see Dashy did have some auth issues with other projects, and it was fixed earlier this year. Have you tried updating to the latest release of Dashy? https://github.com/Lissy93/dashy/issues?q=auth+is%3Aclosed, specifically with this fix (that a lot of the issues link to): https://github.com/Lissy93/dashy/pull/1576

Thanks! And good luck!

**Comment by aleeraser** [comment]
I just faced the very same problem, so I'm hijacking this issue to add more information about my context, hoping it can provide some help.

Same as OP, Cockpit + Firefox is the only combination having this issue. I can use Chrome/Edge and the issue is not present. Also, no issues whatsoever with the other services I host, regardless of the web browser.

I'm using Nginx Proxy Manager as a reverse proxy.
Since yesterday, my setup was the following:
- Cloudflare acts as DNS proxy to NPM with a "Full (Strict)" encryption mode
- opening the proxied NPM page and clicking on the cockpit proxy host would open the login page and everything would be working fine. I'm mentioning NPM here as I'm temporarily using it as a "Homepage" page, before adopting a more appropriate one. Therefore, the issue is not with Dashy, which was mentioned before.

But. Yesterday I set up some DNS rules in Pi-Hole, so that when I'm at home I can access my (sub)domains without going back and forth from Cloudflare.  
Initially everything broke, and Firefox (and Firefox only) presented me `SSL_ERROR_UNRECOGNIZED_NAME_ALERT` randomly when accessing my subdomains (sometimes it would work, sometimes not).  
Luckily, I found [this reddit post](https://www.reddit.com/r/nginxproxymanager/comments/1ffs0sm/domain_doesnt_load_only_on_firefox/) mentioning the same issue, and adding an exception to Firefox secure DNS feature did solve the problem. Everything started working again, except for cockpit.

I can also confirm that hitting enter in the address bar after opening it make the login work correctly.

But. I do see this problem ONLY when browsing NPM from my local ip address (e.g. 192.168.x.x) and clicking on cockpit proxy host. If running from e.g. npm.domain.com (with a local DNS record in Pi-Hole, or with a DNS proxy in Cloudflare), the issue is not there.

**Comment by garrett** [comment]
@aleeraser This is probably the same issue: https://support.mozilla.org/en-US/questions/1132690

> Other browsers may "fill in the gap" by searching for the missing certificate, but Firefox needs to receive it from your server -- or have saved it from at least one previous server -- in order to have a complete chain of trust. That's the problem.

(The link above has some more details too.)

The problem is a server or proxy misconfiguration (not a Cockpit specific issue), and other browsers guess whereas Firefox takes exactly what was given.

**Comment by ninjadev64** [comment]
I've just experienced this too, and came across this by some stroke of luck. Thought it was a problem with putting Cockpit behind Cloudflare Access, but it was just a problem with opening it from the dashboard.

**Comment by saesh** [comment]
Having the same issue with accessing Cockpit through a link from [Homer](https://github.com/bastienwirtz/homer) in a new tab. Directly accessing Cockpit works.

**Comment by martinpitt** [comment]
This all sounds like your reverse proxy doesn't properly handle Websocket connections. In many web servers/proxies this needs to be explicitly enabled.

**Comment by luispabon** [comment]
Why would a reverse proxy matter here and more importantly, work selectively depending on if you clicked a link to get to cockpit?

In any case this is happening to me on a homepage plugin on firefox linking directly to `cockpit:9090`.

**Comment by martinpitt** [comment]
> Why would a reverse proxy matter here and more importantly

It matters for both [nginx](https://github.com/cockpit-project/cockpit/wiki/Proxying-Cockpit-over-NGINX) and [apache](https://github.com/cockpit-project/cockpit/wiki/Proxying-Cockpit-over-Apache-with-LetsEncrypt), so it's highly likely that other reverse proxies need to be configured for this as well.

> In any case this is happening to me on a homepage plugin on firefox linking directly 

ack -- these are multiple different issues then. For everyone running a reverse proxy but direct link works (like reported by @saesh, @nekename , @aleeraser , and @Br0kenSilos -- this here is *NOT* your issue. Please configure your proxy accordingly, and report a new issue if it doesn't work. Thanks!

For @luispabon and the original reporter @Aerowinder this is something different, possibly the [strict cookie setting discussed above](https://github.com/cockpit-project/cockpit/issues/17936#issuecomment-1350595491).

**Comment by luispabon** [comment]
There's a difference on the login request on both scenarios: when clicking a link through to cockpit, the `Authorization` header is missing. Specifically, the following headers:
 *  `Priority`
 * `'Authorization`
 * `X-Superuser`

```diff
+++curl 'https://vault-storage:9090/cockpit/login' -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:137.0) Gecko/20100101 Firefox/137.0' -H 'Accept: */*' -H 'Accept-Language: en-GB,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br, zstd' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Cookie: cockpit=deleted' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin'
---curl 'https://vault-storage:9090/cockpit/login' -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:137.0) Gecko/20100101 Firefox/137.0' -H 'Accept: */*' -H 'Accept-Language: en-GB,en;q=0.5' -H 'Accept-Encoding: gzip, deflate, br, zstd' -H 'DNT: 1' -H 'Connection: keep-alive' -H 'Cookie: cockpit=deleted' -H 'Sec-Fetch-Dest: empty' -H 'Sec-Fetch-Mode: cors' -H 'Sec-Fetch-Site: same-origin' -H 'Priority: u=0' -H 'Authorization: Basic MY_BASE64_CREDS' -H 'X-Superuser: sudo' 
```

**Comment by saesh** [comment]
> For everyone running a reverse proxy but direct link works (like reported by @saesh, @nekename , @aleeraser , and @Br0kenSilos -- this here is NOT your issue. Please configure your proxy accordingly, and report a new issue if it doesn't work. Thanks!

@martinpitt Just to clarify, I am not accessing cockpit behind a reverse proxy, I am linking directly to https://hostname:9090 from Homer.

**Comment by nekename** [comment]
@martinpitt How would this be a reverse proxy issue if it works when I visit Cockpit directly? This is only an issue when I access Cockpit through the new tab page (or the Zero Trust Dashboard).

**Comment by martinpitt** [comment]
Ack, I misunderstood the description then. This is somehow related to the [SameSite cookie restriction](https://github.com/cockpit-project/cockpit/issues/17936#issuecomment-1350595491), but it's not completely clear what exactly. We are not going to relax the auth cookie policy, but this is still not understood -- sure, coming to cockpit from another site should not provide the cookie. But going from the login page to itself ought to work.

(FTR: I will not work on this, at least not this quarter. Help appreciated).

**Comment by tomsik-radek** [comment]
Hi, just adding my 2 cents, I'm also having the same issue. Firefox 148, Flame-Dashboard as my dashboard, no reverse proxy (but they are on two different domains). Using a regular browser bookmark works just fine. I assume SameSite doesn't like me using two different domains for my dashboard and servers? (Imagine dashboard.domain.com and server.domain.eu) Is there a way to set my dashboard domain FQDN as an exception or something?
Or honestly, if this is intended behavior (which I would actually kinda agree with), there should at least be an error message that mentions this.

**Comment by ieris19** [comment]
Any updates on this issue? I am still experiencing this on Fedora 44, versions lists on DNF as `362-1.fc44` which should be latest.

**Comment by he-pennypacker** [comment]
I found this issue and am keen to add some further context to help with getting it fixed.

Using Firefox on CachyOS I am also using a home network dashboard, in this case Heimdall with a link to Cockpit. This is just a simple static page with links - no reverse proxy involved here. Heimdall is served by a different host to the Cockpit host I am trying to log in to.

Cockpit is hosted on a raspberrypi (raspbian). I was using the default repo (Cockpit v 337) and uninstalled and used the backport version (v 362) but unfortunately the issue persists.

If I enter the Cockpit URL directly in a Firefox tab, I can authenticate no issue.

Using Chromium and the Heimdall dashboard link, I can also authenticate no issue.

Using Firefox, if I use the link from the Heimdall dashboard, I just get the endless login loop described by OP.

Let me know if there's any logs etc I can supply to help.

**Comment by zeknoss** [comment]
I can also confirm that manually entering the URL and logging in works while visiting the URL from another page (i.e. gethomepage bookmark) triggers 401 issue.
I use the project behind NginxProxyManager btw.

**Comment by adrianlimmi** [comment]
> I can also confirm that manually entering the URL and logging in works while visiting the URL from another page (i.e. gethomepage bookmark) triggers 401 issue. I use the project behind NginxProxyManager btw.

Can you tell me how your NPM configuration looks like? I can access the login page, but after that I'm still in a loop.

I use the following:

##### /etc/cockpit/cockpit.conf
```
[WebService]
Origins = https://<domain> wss://<domain>
ProtocolHeader = X-Forwarded-Proto
```

##### NPM "Advanced config"
```
# Required to proxy the connection to Cockpit
proxy_set_header Host $host;
proxy_set_header X-Forwarded-Proto $scheme;

# Required for web sockets to function
proxy_buffering off;
proxy_set_header Upgrade $http_upgrade;
proxy_set_header Connection "upgrade";
```

**Comment by zeknoss** [comment]
> > I can also confirm that manually entering the URL and logging in works while visiting the URL from another page (i.e. gethomepage bookmark) triggers 401 issue. I use the project behind NginxProxyManager btw.
> 
> Can you tell me how your NPM configuration looks like? I can access the login page, but after that I'm still in a loop.
> 
> I use the following:
> ##### /etc/cockpit/cockpit.conf
> 
> ```
> [WebService]
> Origins = https://<domain> wss://<domain>
> ProtocolHeader = X-Forwarded-Proto
> ```
> 
> ##### NPM "Advanced config"
> 
> ```
> # Required to proxy the connection to Cockpit
> proxy_set_header Host $host;
> proxy_set_header X-Forwarded-Proto $scheme;
> 
> # Required for web sockets to function
> proxy_buffering off;
> proxy_set_header Upgrade $http_upgrade;
> proxy_set_header Connection "upgrade";
> ```

Hey @adrianlimmi ,
Here's my advanced config:
```
location / {
    include /snippets/proxy.conf;
    proxy_pass $forward_scheme://$server:$port;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection "upgrade";
}
```


proxy.conf:
```
## Headers
proxy_set_header Host $host;
proxy_set_header X-Original-URL $scheme://$http_host$request_uri;
proxy_set_header X-Forwarded-Proto $scheme;
proxy_set_header X-Forwarded-Host $http_host;
proxy_set_header X-Forwarded-URI $request_uri;
proxy_set_header X-Forwarded-Ssl on;
proxy_set_header X-Forwarded-For $remote_addr;
proxy_set_header X-Real-IP $remote_addr;
#proxy_set_header Upgrade $http_upgrade;
#proxy_set_header Connection "upgrade";

## Basic Proxy Configuration
client_body_buffer_size 128k;
proxy_next_upstream error timeout invalid_header http_500 http_502 http_503; ## Timeout if the real server is dead.
proxy_redirect  http://  $scheme://;
proxy_http_version 1.1;
proxy_cache_bypass $cookie_session;
proxy_no_cache $cookie_session;
proxy_buffers 64 256k;

## Trusted Proxies Configuration
## Please read the following documentation before configuring this:
##     https://www.authelia.com/integration/proxies/nginx/#trusted-proxies-and-integration-security
# set_real_ip_from 10.0.0.0/8;
# set_real_ip_from 172.16.0.0/12;
# set_real_ip_from 192.168.0.0/16;
# set_real_ip_from fc00::/7;
real_ip_header X-Forwarded-For;
real_ip_recursive on;

## Advanced Proxy Configuration
send_timeout 5m;
proxy_read_timeout 360;
proxy_send_timeout 360;
proxy_connect_timeout 360;
```
